c8d/delete: Add support for deleting specific platforms

vvoland opened the pull request:
- addresses: https://github.com/moby/moby/issues/47680

**- What I did**

**- How I did it**

**- How to verify it**

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
`DELETE /images/{name}` now supports a `platforms` query parameter. It accepts an array of JSON-encoded OCI Platform objects, allowing for selecting a specific platforms to delete content for.
```

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
daemon/images: Make ImageDelete take opts struct

Pass whole `RemoveOptions` struct instead of individual `force` and
`prune` parameters.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/delete: Extract untagReferences

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/delete: Support deleting specific platforms

This change adds the ability to delete a specific platform from a
multi-platform image.

Previously, image deletion was an all-or-nothing operation - when
deleting a multi-platform image, all platforms would be removed
together. This change allows users to selectively remove individual
platforms from a multi-architecture image while keeping other platforms
intact.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d: Extract memoryLabelStore

Move to an internal testutils package

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on api/server/httputils/form.go:
```diff
@@ -162,3 +162,22 @@ func DecodePlatform(platformJSON string) (*ocispec.Platform, error) {
+// DecodePlatforms decodes the OCI platform JSON string into a Platform struct.
+//
+// Typically, the argument is a value of: r.Form["platform"]
+func DecodePlatforms(platformJSONs []string) ([]ocispec.Platform, error) {
```
Heh, I thought we already had this, but it really now that for save/load we didn't yet implement multiple platforms, so we currently still produce an error; https://github.com/moby/moby/blob/7937f0846c130e91836f5f640c41ce12d07da4c8/api/server/router/image/image_routes.go#L279-L291

thaJeztah commented on daemon/containerd/image_delete.go:
```diff
@@ -411,6 +385,25 @@ func (*imageDeleteConflict) Conflict() {}
 // nil if there are none. It takes a bitmask representing a
 // filter for which conflict types the caller cares about,
 // and will only check for these conflict types.
+
+// untagReferences deletes the given image references and returns the appropriate response records
+func (i *ImageService) untagReferences(ctx context.Context, refs []c8dimages.Image) ([]imagetypes.DeleteResponse, error) {
```
Looks like the GoDoc for `checkImageDeleteConflict` got separated from the function

vvoland replied:
whoops, thanks!

thaJeztah commented on internal/c8dutil/memory_label_store.go:
```diff
@@ -0,0 +1,58 @@
+
+func NewMemoryLabelStore() *memoryLabelStore {
+	return &memoryLabelStore{
+		labels: make(map[digest.Digest]map[string]string),
```
Looks like the implementation didn't change, so is still handling nil values for labels.

So you could either;


- Remove the constructor, and export `memoryLabelStore`
- Remove this line, and just return ` memoryLabelStore{}`
- Change the implementations to remove the `if s.labels == nil` checks

vvoland replied:
Done

thaJeztah commented on daemon/containerd/fake_service_test.go:
```diff
@@ -244,48 +243,3 @@ func (s *delayedStore) Update(ctx context.Context, info content.Info, fieldpaths
 	return s.store.Update(ctx, info, fieldpaths...)
 }
-
-type memoryLabelStore struct {
```
Looks like this was a copy of an implementation in `distribution` - should we remove that one as well? https://github.com/moby/moby/blob/8601b22f5db511354d643a7722d11d33aa7ae13f/distribution/manifest_test.go#L46-L68

thaJeztah commented on internal/c8dutil/memory_label_store.go:
```diff
@@ -0,0 +1,58 @@
+package c8dutil
```
Perhaps make this its own package to prevent `c8dutil` from becoming too much of a grab-bag of things (with many imports); given that this is a test-util, perhaps `internal/testutils/labelstore` (or `internal/testutils/memorylabelstore`, but that's a bit lengthy)

thaJeztah commented on daemon/containerd/image_delete.go:
```diff
@@ -52,14 +52,17 @@ import (
 //
 // TODO(thaJeztah): image delete should send prometheus counters; see https://github.com/moby/moby/issues/45268
-func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) (response []imagetypes.DeleteResponse, retErr error) {
+func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, options imagetypes.RemoveOptions) (response []imagetypes.DeleteResponse, retErr error) {
```
Looks like the GoDoc needs updating as it's still referencing the `prune` and `force` arguments; while updating, perhaps also remove use of future tense ("will each attempt" -> "are attempted to be deleted quietly")

```go
// cannot be removed unless either there are other repository references to the
// same image or force is true. Following removal of the repository reference,
```

```go
// If prune is true, ancestor images will each attempt to be deleted quietly,
```

vvoland replied:
I adjusted it a bit

thaJeztah commented on daemon/images/image_prune.go:
```diff
@@ -114,7 +114,10 @@ deleteImagesLoop:
 				for _, ref := range refs {
-					imgDel, err := i.ImageDelete(ctx, ref.String(), false, true)
+					imgDel, err := i.ImageDelete(ctx, ref.String(), imagetypes.RemoveOptions{
+						Force:         false,
```
Could be removed (as default value)

vvoland replied:
Done

thaJeztah commented on daemon/images/image_prune.go:
```diff
@@ -123,7 +126,10 @@ deleteImagesLoop:
 			hex := id.Digest().Encoded()
-			imgDel, err := i.ImageDelete(ctx, hex, false, true)
+			imgDel, err := i.ImageDelete(ctx, hex, imagetypes.RemoveOptions{
+				Force:         false,
```
Same here

thaJeztah commented on daemon/containerd/image_delete.go:
```diff
@@ -127,7 +141,16 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, options
 		using := func(c *container.Container) bool {
 			if c.ImageID == imgID {
-				return true
+				if len(options.Platforms) > 0 {
```
Can reverse the condition here and return early;

```go
if len(options.Platforms == 0) {
    return true
}
```

thaJeztah commented on daemon/containerd/image_delete.go:
```diff
@@ -127,7 +141,16 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, options
+						if pm.Match(c.ImagePlatform) {
+							return true
+						}
+					}
```
Perhaps this path could use a comment to outline we're intentionally _not_ returning `false` after this, something like;

```go
// No match for the image reference, but continue to check if used as mounted image
```

(better suggestions welcome!)


(because I was wondering if we forgot a `return false` here; the comment may help future visitors 😄

thaJeztah replied:
☝️ really wondering if this function needs splitting up, because there's a _ton_ of conditions (not for this PR!)

vvoland replied:
Yes actually I started by refactoring it to make the flow more linear, but turned out that I broke some edge cases in the process and didn't want to spend more time on that 🙈

thaJeztah commented on docs/api/version-history.md:
```diff
@@ -21,6 +21,9 @@ keywords: "API, Docker, rcli, REST, documentation"
   device driver.
   Currently only the CDI device driver is supported.
+* `DELETE /images/{name}` now supports a `platforms` query parameter. It accepts
+  an array of JSON-encoded OCI Platform objects, allowing for selecting a specific
+  platforms to delete content for.
```
```suggestion
  an array of JSON-encoded OCI Platform objects, allowing for selecting a specific
  platform to delete content for.
```

thaJeztah replied:
In that case, probably "to select the platforms for which to delete the content for" (or something along those lines)

Currently it say "a specific" (singular) followed by "platforms" plural 😅

thaJeztah commented on daemon/images/image_delete.go:
```diff
@@ -66,7 +67,16 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, options
+	case 1:
+		platform = &options.Platforms[0]
+	default:
+		return nil, errdefs.InvalidParameter(errors.New("multiple platforms are not supported"))
```
This is probably something we want to implement in future? (if so, perhaps we should create a tracking-ticket for that work)

thaJeztah reviewed (changes requested):
Had some comments that I didn't submit yet 😓 (not sure if I fully completed review, but let me post these)

thaJeztah commented on api/server/httputils/form.go:
```diff
@@ -162,3 +162,22 @@ func DecodePlatform(platformJSON string) (*ocispec.Platform, error) {
+//
+// Typically, the argument is a value of: r.Form["platform"]
+func DecodePlatforms(platformJSONs []string) ([]ocispec.Platform, error) {
+	if len(platformJSONs) == 0 {
+		return nil, nil
+	}
```
No need to change; I guess this is a bit redundant (it won't loop if there's nothing to loop, so return `output` with its nil value)

thaJeztah reviewed (approved):
LGTM

dmcgowan commented on daemon/containerd/image_delete.go:
```diff
@@ -194,9 +199,71 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
+	// The same manifest can be referenced by different image indexes.
+	var response []imagetypes.DeleteResponse
+	for _, d := range toDelete {
+		if err := i.content.Delete(ctx, d.Digest); err != nil {
```
We should avoid doing any direct deletion like this, removing root objects and updating reference labels is the preferred solution. There is no other transactional way to ensure nothing is using the content

vvoland replied:
Right, it's not really ideal, but overall we assume that `moby` containerd namespace is exclusively accesses by `dockerd`.

There's a check that at least makes sure that it's not used by any container and that requires an explicit `force`.

Perhaps we should make it ONLY work with `force` now? That would make it clear that the operation is potentially dangerous, while still giving an escape hatch for users that want to get rid of platforms they don't need.

vvoland committed:
c8d/delete: Require --force when deleting platforms

When deleting specific platforms from a multi-platform image, there's
potential risk as platform-specific content might be shared across
images/containers.

For now, require `force` parameter to make the user aware.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah reviewed (approved):
still LGTM
